lib/std/math/powi.zig

```diff
@@ -10,108 +10,92 @@ const testing = std.testing;
+///  - powi(x, y)   = Underflow for y < 0
 pub fn powi(comptime T: type, x: T, y: T) (error{
     Overflow,
     Underflow,
 }!T) {
-    const info = @typeInfo(T);
+    const bit_size = @bitSizeOf(T);
 
     comptime assert(@typeInfo(T) == .Int);
 
-    //  powi(x, +-0)   = 1 for any x
-    if (y == 0 or y == -0) {
-        return 1;
+    // integer value 1 cannot be coerced to type 'i1'
+    const does_one_overflow = bit_size == 0 or T == i1;
+    const is_y_even = !does_one_overflow and y & 1 == 0;
+
+    if (x == 1 or y == 0 or (x == -1 and is_y_even)) {
+        if (does_one_overflow) {
+            return error.Overflow;
+        } else {
+            return 1;
+        }
     }
 
-    switch (x) {
-        //  powi(0, y)     = 0 for any y
-        0 => return 0,
-
-        //  powi(1, y)     = 1 for any y
-        1 => return 1,
-
-        else => {
-            //  powi(x, y)     = Overflow for for y >= @sizeOf(x) - 1 y > 0
-            //  powi(x, y)     = Underflow for for y > @sizeOf(x) - 1 y < 0
-            const bit_size = @sizeOf(T) * 8;
-            if (info.Int.signedness == .signed) {
-                if (x == -1) {
-                    //  powi(-1, y)    = -1 for for y an odd integer
-                    //  powi(-1, y)    = 1 for for y an even integer
-                    if (@mod(y, 2) == 0) {
-                        return 1;
-                    } else {
-                        return -1;
-                    }
-                }
-
-                if (x > 0 and y >= bit_size - 1) {
-                    return error.Overflow;
-                } else if (x < 0 and y > bit_size - 1) {
-                    return error.Underflow;
-                }
-            } else {
-                if (y >= bit_size) {
-                    return error.Overflow;
-                }
-            }
+    if (x == -1) {
+        return -1;
+    }
 
-            var base = x;
-            var exp = y;
-            var acc: T = 1;
-
-            while (exp > 1) {
-                if (exp & 1 == 1) {
-                    if (@mulWithOverflow(T, acc, base, &acc)) {
-                        if (x > 0) {
-                            return error.Overflow;
-                        } else {
-                            return error.Underflow;
-                        }
-                    }
-                }
-
-                exp >>= 1;
-
-                if (@mulWithOverflow(T, base, base, &base)) {
-                    if (x > 0) {
-                        return error.Overflow;
-                    } else {
-                        return error.Underflow;
-                    }
-                }
-            }
+    if (x == 0) {
+        if (y > 0) {
+            return 0;
+        } else {
+            // not overflow in strict sense
+            return error.Overflow;
+        }
+    }
+    // x >= 2 or x <= -2
+    if (y >= bit_size) {
+        return error.Overflow;
+    }
+    if (y < 0) {
```

Comment:
should be checked first to remove the `x==0 and `y<0` branch.

Reply:
`0^(-1)` should be an overflow, not an underflow?

Reply:
Julia thinks `0^(-1)` is Inf.

Reply:
which is overflow per the code's definition? I don't see how that would remove branch, as `x==0 and y<0` and `x !=0 and y<0` (and implictly `x != -1 and x !=1`, because it is already checked) should give different result (Overflow and Underflow respectively)

Reply:
correct, my bad then.